docs/design/2022-12-07-keyspace.md

```diff
@@ -0,0 +1,210 @@
+![keyspace-state-machine.png](imgs%2Fkeyspace-state-machine.png)
+
+### TiDB
+
+#### Using API V2 and etcd namespace to isolate TiKV data and etcd path by Keyspace.
+
+1. Configure the keyspace name by setting `keyspace-name` in the configuration file. 
+   * When the TiDB server starts, it will open a TiKV driver. The driver has a TxnKV client object called `tikv.KVStore`. If `keyspace-name` is set, it will use `tikv.KVStore` via API V2 to access TiKV.
+2. The etcd path should have a different prefix based on the etcd namespace. 
+   * An etcd client is created during domain initialization. If `keyspace-name` is set, it will add an etcd namespace when creating the etcd client. 
+   * The format of the etcd namespace is `/keyspaces/tidb/$keyspaceId`.
+
+#### TiDB BR
+
+If "keyspace-name" is set when executing the BR command, it supports restore and backup for the specified keyspace data. The BR task will parse the old keyspace prefix from the source data, and get target keyspace from KVStore.codec. The old keyspace prefix and new keyspace prefix are used to update the old key prefix and new key prefix of rewrite rules. Restore checksum also requires the old and new keyspace prefixes. The backup task needs to add a target keyspace prefix when building ranges in `BackupRequest` to scan for the data of the specified keyspace in TiKV.
+
+#### TiDB Lightning
+
+If `keyspace-name` is set when executing the `Lightning` command, it will encode the key to add the keyspace prefix before appending rows to the SST file, and decode the key before recording conflict data into the TiDB table.
+
+#### TiCDC
+
+If `keyspace-name` is set when creating a changefeed to the TiCDC server, the server will only capture the table data within that keyspace for the changefeed. For changefeeds configured with a keyspace, the TiCDC server will encode the table range with the keyspace prefix to subscribe to the table data from the TiKV server, and it will also decode the keyspace prefix from the key received from the TiKV server to retrieve the table data. If `keyspace-name` is not set, the TiCDC server will not capture the table data from [API V2](https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md#tikv-server).
+
+#### TiDB GC
+
+##### Concepts of GC management type:
+
+1. Global GC:
+   - Represents the previous default GC logic. There is a TiDB GC Worker that is responsible for calculating the global GC safe point for the entire cluster.
+   - The default GC management type for a keyspace is global GC.
+2. Keyspace Level GC:
+   - Indicates that the keyspace will advance its own GC safe point.
+   - `Keyspace` GC related data includes: minimum start timestamp, GC safe point, and service safe point, which are stored in the dedicated path of each keyspace in PD.
```

Comment:
Can the GC parameters (e.g. tidb_gc_life_time) by set by each tenant separately? If so, by what kind of interface? HTTP/RPC?

Reply:
> Can the GC parameters (e.g. tidb_gc_life_time) by set by each tenant separately? If so, by what kind of interface? HTTP/RPC?

Keyspace users can set tidb_gc_life_time directly using the SET SQL command just as before. This is because the system variables for each keyspace user are isolated.

Reply:
In the keyspace scenario, all SQL operations remain the same as before.